azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/AuxiliaryTokensInterceptor.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.management.resources.fluentcore.utils;
+
+import com.microsoft.azure.credentials.AzureTokenCredentials;
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+
+import java.io.IOException;
+
+/**
+ * An interceptor for cross-tenant authorization in Azure.
+ */
+public final class AuxiliaryTokensInterceptor implements Interceptor {
+
+    private static final String AUTHORIZATION_AUXILIARY_HEADER = "x-ms-authorization-auxiliary";
+    private static final String SCHEMA = "Bearer";
+
+    private final AzureTokenCredentials[] tokens;
+
+    /**
+     * Initialize an auxiliary interceptor with the list of AzureTokenCredentials.
+     *
+     * @param tokens the AzureTokenCredentials list
+     */
+    public AuxiliaryTokensInterceptor(AzureTokenCredentials... tokens) {
+        this.tokens = tokens;
+    }
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        if (this.tokens == null || this.tokens.length == 0) {

```

Comment:
Can we try `chain.proceed(chain.request()` first and only authenticate the auxiliary credentials if a certain error comes back? This will avoid unnecessary network calls for method calls not requiring auxiliary tokens.

Reply:
Fixed. LinkedAuthorizationFailed will throw at the first spot.